subprojects/core/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathFingerprintingStrategy.java

```diff
@@ -150,13 +151,15 @@ public boolean preVisitDirectory(DirectorySnapshot directorySnapshot) {
 
         @Override
         public void visitFile(FileSystemLocationSnapshot fileSnapshot) {
-            if (fileSnapshot.getType() == FileType.RegularFile) {
+            if (fileSnapshot instanceof RegularFileSnapshot) {
                 HashCode normalizedContent = fingerprintFile((RegularFileSnapshot) fileSnapshot);
                 if (normalizedContent != null) {
                     delegate.visit(fileSnapshot, normalizedContent);
                 }
-            } else if (!relativePathSegmentsTracker.isRoot()) {
-                throw new GradleException(String.format("Couldn't follow symbolic link '%s'.", fileSnapshot.getAbsolutePath()));
+            } else if (fileSnapshot instanceof MissingFileSnapshot) {
+                if (!relativePathSegmentsTracker.isRoot() && ((MissingFileSnapshot) fileSnapshot).isSymbolicLink()) {
```

Comment:
calling `isSymbolicLink` only if every other check fails, since it's expensive

Reply:
Why don't we fail if the missing file is not a symbolic link? If it is not a symbolic link, then it is an unreadable file. We should fail in that case as well.

Reply:
Reworded the error

Reply:
Reverted